events: Update docs for beta release (#23036)
hashicorp/vault: website/content/docs/commands/events.mdx

---
layout: docs
page_title: events - Command
description: |-
  The "events" command interacts with the Vault events subsystem.
---

# events

The `events` command is used to subscribe to Vault

> [events](/vault/docs/concepts/events) via `events subscribe`. It shows a

Reviewer: Follow-up question: what does "it shows a real-time display of events" mean? Is there a separate `events` subcommand that lists available events or do we just return a list of active events as part of a successful subscription call?

Author: There is no subcommand for showing available events. This returns an on-going list of all events that are happening, and will continue forever (as opposed to almost every other Vault command, which do one thing, return a response, and are done).

> The `events` command is used to subscribe to Vault
> [events](/vault/docs/concepts/events) via `events subscribe`. It shows a
> real-time display of events generated by Vault on standard out.

Reviewer: ```suggestion
Use the `events` command to get a real-time display of
[events](/vault/docs/concepts/events) generated by Vault and subscribe to Vault
events. Note that the `events subscribe` runs indefinitly and will not exit on
its own unless it encounters an unexpected error. Similar to `tail -f` in the
Unix world, you must terminate the process from the command line to end the
`events` command.
```

Style correction: write in active voice

> The desired event types (sometimes called "topics") must be specified, which
> may be a glob pattern (with `*` treated as a wildcard).
> 
> The output will be JSON objects serialized using the default protobuf JSON
> serialization format, with one line per event received.

Reviewer: ```suggestion
Specify the desired event types (also called "topics") as a glob pattern. To match
against multiple events, use `*` as a wildcard. The command returns serialized
JSON objects in the default protobuf JSON serialization format with one line per
event received.
```

Style correction: write in active voice

> Note that this command is designed to run indefinitely. It is similar to
> `tail -f` in the Unix world. This command will not exit on its own unless it
> encounters an unexpected error. As a user, you must terminate this process
> yourself to shut it down.

Reviewer: ```suggestion
```

Suggest including the behavioral note with the intro sentence so folks have a clear idea of how the command behaves right from the jump.

## Examples

Subscribe to all events:

```shell-session
$ vault events subscribe '*'
```

Subscribe to all KV events:

```shell-session
$ vault events subscribe 'kv*'
```

Subscribe to all `kv-v2/data-write` events:

```shell-session
$ vault events subscribe kv-v2/data-write
```

> Subscribe to all non-`data-write` KV events for the secret `secret/data/foo` in
> the current namespace as well as namespace `ns1`:

Reviewer: ```suggestion
Subscribe to all KV events in the current and `ns1` namespaces for the secret `secret/data/foo` that do not involve writing data:
```

Suggest restructuring the sentence to make it a little easier to parse.

```shell-session
$ vault events subscribe -namespaces=ns1 -filter='data_path == secret/data/foo and operation != "data-write"' 'kv*'
```

## Usage

> The following flags are available in addition to the [standard set of
> flags](/vault/docs/commands) included on all commands.

Reviewer: ```suggestion
`events subscribe` supports the following flags in addition to the [standard set of
flags](/vault/docs/commands) included on all commands.
```

Style correction: write in active voice

  specified duration.

- `-filter` `(string: "")` - Filter expression used to select events to be sent
  through the WebSocket.

>   See [Boundary's docs](/boundary/docs/concepts/filtering) on how these
>   expressions are evaluated and what options are available.

Reviewer: ```suggestion
Refer to the [Filter expressions](/boundary/docs/concepts/filtering) guide
in the Boundary documentation for a complete list of filtering options and an
explanation on how we evaluate filter expressions.
```

Style correction: avoid general anchor text

  through the WebSocket.

  See [Boundary's docs](/boundary/docs/concepts/filtering) on how these

>   expressions are evaluated and what options are available.

Reviewer: Why are we sending folks to the Boundary docset?

Author: That page is independent of Boundary. They just happen to have written the best documentation on the language implemented by hashicorp/go-bexpr, and there's no sense duplicating it.

  The following values are available in the filter expression:
  - `event_type`: the event type, e.g., `kv-v2/data-write`.
	- `operation`: the operation name that caused the event, e.g., `write`.
	- `source_plugin_mount`: the mount of the plugin that produced the event,
    e.g., `secret/`
	- `data_path`: the API path that can be used to access the data of the secret related to the event, e.g., `secret/data/foo`
	- `namespace`: the path of the namespace that created the event, e.g., `ns1/`

>   By default, the filter will be empty and will match all events that the
>   request has access to for the given event types specified in the request.
> 
>   A filter could be as simple as `data_path == secret/data/foo` to specify that
>   only events that refer to the `data_path` of `secret/data/foo` should be sent
>   via the WebSocket.
> 
>   A filter can also be more complex, for example:

Reviewer: ```suggestion
The filter string is empty by default. Unfiltered subscription requests match to
all events that the requestor has access to for the target event type. When the
filter string is not empty, Vault applies the filter conditions after the policy
checks to narrow the events provided in the response.

Filters can be straightforward path matches like `data_path == secret/data/foo`,
which specifies that Vault should pass return events that refer to the
`secret/data/foo` secret to the WebSocket. Or more complex statements that
exclude specific operations. For example:
```
Style correction: write in active voice, avoid implying something is/isn't "simple"

>   ```

Reviewer: I think the code fence should be typed as `shell-session` here.

Author: It's not a shell-session though. It's a generic boolean expression in the hashicorp/go-bexpr language.

  data_path == secret/data/foo and operation != write
  ```

>   The filter is applied after policy checks.

Reviewer: ```suggestion
```

Folded into the suggestion above

> ### Enterprise Options

Reviewer: ```suggestion
### Enterprise options
```

Style correction: use sentence case for headings

<EnterpriseAlert product="vault" />

> - `-namespaces` `(string)` - Additional child namespaces to subscribe to. Can
>   be repeated to add more namespace patterns. The namespace of the request is
>   automatically prepended, so specifying `ns2` when the request is in the `ns1`
>   namespace will result in subscribing to `ns1/ns2` (in addition to `ns1`).
>   Patterns can include `*` characters to indicate wildcards. The default is to
>   subscribe only to the request's namespace.

Reviewer: ```suggestion
- `-namespaces` `(string)` - Additional **child** namespaces for the
  subscription. Repeat the flag to add additional namespace patterns to the
  subscription request. Vault automatically prepends the issuing namespace for
  the request to the provided namespace. For example, if you include
  `-namespaces=ns2` on a request made in the `ns1` namespace, Vault will attempt
  to subscribe you to events under the `ns1/ns2` and `ns1` namespaces. You can
  use the `*` character to include wildcards in the namespace pattern. By
  default, Vault will only subscribe to events in the requesting namespace.
```

Style correction: write in active voice with one idea per sentence

>   -> Note: A root token, or a token associated with entities and groups with
>      the appropriate policies, must be used for the event subscription
>      request to be able to subscribe to multiple namespaces. See
>      [this tutorial](/vault/tutorials/enterprise/namespaces) for an example
>      setup.

Reviewer: ```suggestion
<Note>

  To subscribe to events across multiple namespaces, you must provide a root
  token or a token associated with appropriate policies across all the targeted
  namespaces. Refer to the
  [Secure multi-tenancy with namespaces](/vault/tutorials/enterprise/namespaces)
  tutorial for configuring your Vault instance appropriately.

</Note>
```

Style correction: use new notation for asides, write in active voice, avoid generic anchor text
